include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -1545,20 +1545,22 @@ __pattern_rotate(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator __new_f
     auto __buf = __keep(__first, __last);
     auto __temp_buf = oneapi::dpl::__par_backend_hetero::__buffer<_ExecutionPolicy, _Tp>(__exec, __n);
 
-    auto __temp_rng =
+    auto __temp_rng_w =
         oneapi::dpl::__ranges::all_view<_Tp, __par_backend_hetero::access_mode::write>(__temp_buf.get_buffer());
 
     const auto __shift = __new_first - __first;
     oneapi::dpl::__par_backend_hetero::__parallel_for(
         oneapi::dpl::__par_backend_hetero::make_wrapped_policy<__rotate_wrapper>(__exec),
         unseq_backend::__rotate_copy<typename ::std::iterator_traits<_Iterator>::difference_type>{__n, __shift}, __n,
-        __buf.all_view(), __temp_rng);
```

Comment:
This falls in a similar category as the `partial_sort_copy` case above.

For buffers, I believe the implicit data dependencies of the accessor to `__temp_buf` should order these appropriately, and the wait on the last `__parallel_for` should remove any race with the return.

However, considering USM pointer inputs I think we lose those and this can be a bug that requries the `wait()`

Reply:
Yes, it is not a great situation, but you are correct because `__temp_buf` is a temporary `sycl::buffer` these waits are unnecessary.  Assuming we remove both `wait()`,  we should add comments for each to describe the implicit synchronization as it is not clear. (I missed it even with this topic in my mind when looking back at this PR)